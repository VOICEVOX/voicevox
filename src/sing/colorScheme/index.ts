import Color from "colorjs.io";

import {
  ColorSchemeConfig,
  ColorScheme,
  ColorSpace,
  ColorName,
  ColorPalette,
  CSSColorString,
  OklchColorArray,
  ColorType,
  ColorSet,
} from "./types";

import { colorRoles } from "./constants";
import { defaultGenerator } from "./generator";

// 色変換関数
export const toOklchArray = (colorString: CSSColorString): OklchColorArray => {
  try {
    const color = new Color(colorString);
    const oklch = color.to("oklch");
    return [oklch.l, oklch.c, oklch.h];
  } catch (error) {
    throw new Error(`Failed to convert color to OKLCH: ${error}`);
  }
};

export const fromOklchArray = (oklch: OklchColorArray): CSSColorString => {
  return new Color("oklch", oklch).toString();
};

export const toCssString = (
  colorArray: OklchColorArray,
  format: ColorSpace = "oklch",
): CSSColorString => {
  try {
    const color = new Color("oklch", colorArray);
    switch (format) {
      case "oklch":
        return color.toString({ format: "oklch" });
      case "hex":
        return color.to("srgb").toString({ format: "hex" });
      case "rgb":
        return color.to("srgb").toString();
      case "p3":
        return color.to("p3").toString();
      default:
        throw new Error(`Unsupported color format: ${format}`);
    }
  } catch (error) {
    throw new Error(`Failed to convert color to CSS string: ${error}`);
  }
};

// 色調整関数
export const adjustLightness = (
  color: OklchColorArray,
  delta: number,
): OklchColorArray => {
  const [l, c, h] = color;
  return [l + delta, c, h];
};

export const adjustChroma = (
  color: OklchColorArray,
  delta: number,
): OklchColorArray => {
  const [l, c, h] = color;
  return [l, c + delta, h];
};

export const adjustHue = (
  color: OklchColorArray,
  delta: number,
): OklchColorArray => {
  const [l, c, h] = color;
  return [l, c, (h + delta) % 360];
};

// パレットの生成
export const createPalette = (
  baseColor: OklchColorArray,
  type: ColorType,
): ColorPalette => {
  try {
    return {
      name: type,
      colors: defaultGenerator(baseColor, type),
    };
  } catch (error) {
    throw new Error(`Failed to create palette for ${type}: ${error}`);
  }
};

// パレットから明度にあわせた色を取得
const getColorFromPalette = (
  palette: ColorPalette,
  lightness: number,
): string => {
  try {
    const closestKey = Object.keys(palette.colors)
      .map(Number)
      .reduce((a, b) =>
        Math.abs(b - lightness) < Math.abs(a - lightness) ? b : a,
      );
    return palette.colors[closestKey];
  } catch (error) {
    throw new Error(`Failed to get color from palette: ${error}`);
  }
};

// ColorSchemeを生成
// NOTE: 大きいので分割
export const createColorScheme = (config: ColorSchemeConfig): ColorScheme => {
  try {
    // パレットの生成
    const createPaletteFromConfig = (
      colorType: ColorType,
      colorString?: string,
    ): ColorPalette => {
      const baseColor = colorString
        ? toOklchArray(colorString)
        : toOklchArray(config.primary);
      return createPalette(baseColor, colorType);
    };

    const palettes: Record<ColorType, ColorPalette> = {
      primary: createPaletteFromConfig("primary", config.primary),
      secondary: createPaletteFromConfig("secondary", config.secondary),
      tertiary: createPaletteFromConfig("tertiary", config.tertiary),
      neutral: createPaletteFromConfig("neutral", config.neutral),
      neutralVariant: createPaletteFromConfig(
        "neutralVariant",
        config.neutralVariant,
      ),
      error: createPaletteFromConfig("error", config.error),
    };

    // ロールに基づいた色の生成
    const roles = colorRoles.reduce(
      (acc, role) => {
        const palette = palettes[role.palette];
        if (!palette) {
          throw new Error(`Palette not found for role: ${role.name}`);
        }
        return {
          ...acc,
          [role.name]: {
            light: getColorFromPalette(palette, role.light),
            dark: getColorFromPalette(palette, role.dark),
          },
        };
      },
      {} as Record<ColorName, ColorSet>,
    );

    // エイリアスカラーの生成
    // (eg. buttonBackground: primaryかつ明度0.1などパレットから明度取得)
    const aliasColors = (config.aliasColors || []).reduce(
      (acc, aliasColor) => {
        const palette = palettes[aliasColor.palette];
        if (!palette) {
          throw new Error(`Palette not found for role: ${aliasColor.palette}`);
        }
        return {
          ...acc,
          [aliasColor.name]: {
            light: getColorFromPalette(palette, aliasColor.light),
            dark: getColorFromPalette(palette, aliasColor.dark),
          },
        };
      },
      {} as Record<ColorName, ColorSet>,
    );

    /*
    // カスタムカラー
    // (eg. customButtonBackground: secondaryと同じロジック+Hueが違うなど / onCustomButtonなど派生ロール)
    // カスタムカラーの生成
    const customColors = (config.customColors || []).reduce(
      (acc, customColor) => {
        const customPalette = createPalette(
          toOklchArray(customColor.sourceColor),
          customColor.palette,
        );
        const baseRole = colorRoles.find(
          (role) => role.name === customColor.palette,
        );
        if (!baseRole) {
          throw new Error(
            `Base role not found for custom color: ${customColor.name}`,
          );
        }

        const mainColor = {
          [customColor.name]: {
            source: customColor.sourceColor,
            light: getColorFromPalette(customPalette, baseRole.light),
            dark: getColorFromPalette(customPalette, baseRole.dark),
          },
        };

        const derivedRoles = [
          { name: `on${customColor.name}`, light: 1, dark: 0.3 },
          { name: `${customColor.name}Container`, light: 0.9, dark: 0.4 },
          { name: `on${customColor.name}Container`, light: 0.2, dark: 0.9 },
        ];

        const derivedColors = derivedRoles.reduce(
          (derivedAcc, role) => {
            derivedAcc[role.name] = {
              light: getColorFromPalette(customPalette, role.light),
              dark: getColorFromPalette(customPalette, role.dark),
            };
            return derivedAcc;
          },
          {} as Record<ColorName, { light: string; dark: string }>,
        );

        return {
          ...acc,
          ...mainColor,
          ...derivedColors,
        };
      },
      {} as Record<string, { light: string; dark: string }>,
    ); */

    return {
      name: config.name,
      displayName: config.displayName,
      palettes,
      roles,
      aliasColors,
      customColors: {},
      config,
    };
  } catch (error) {
    throw new Error(`Failed to create color scheme: ${error}`);
  }
};

export const cssVariablesFromColorScheme = (
  colorScheme: ColorScheme,
): Record<string, string> => {
  const cssVars: Record<string, string> = {};

  const toKebabCase = (str: string) => {
    return str.replace(/([a-z])([A-Z])/g, "$1-$2").toLowerCase();
  };

  // roles の CSS 変数
  Object.entries(colorScheme.roles).forEach(([key, value]) => {
    cssVars[`--md-sys-color-${toKebabCase(key)}-light`] = value.light;
    cssVars[`--md-sys-color-${toKebabCase(key)}-dark`] = value.dark;
  });

  // aliasColors の CSS 変数
  Object.entries(colorScheme.aliasColors).forEach(([key, value]) => {
    cssVars[`--md-custom-color-${toKebabCase(key)}-light`] = value.light;
    cssVars[`--md-custom-color-${toKebabCase(key)}-dark`] = value.dark;
  });

  // customColors の CSS 変数
  Object.entries(colorScheme.customColors).forEach(([key, value]) => {
    cssVars[`--md-custom-color-${toKebabCase(key)}-light`] = value.light;
    cssVars[`--md-custom-color-${toKebabCase(key)}-dark`] = value.dark;
  });

  return cssVars;
};
