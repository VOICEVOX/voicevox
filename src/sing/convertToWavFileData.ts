export const convertToWavFileData = (audioBuffer: AudioBuffer) => {
  const bytesPerSample = 4; // Float32
  const formatCode = 3; // WAVE_FORMAT_IEEE_FLOAT

  const numberOfChannels = audioBuffer.numberOfChannels;
  const numberOfSamples = audioBuffer.length;
  const sampleRate = audioBuffer.sampleRate;
  const byteRate = sampleRate * numberOfChannels * bytesPerSample;
  const blockSize = numberOfChannels * bytesPerSample;
  const dataSize = numberOfSamples * numberOfChannels * bytesPerSample;

  const buffer = new ArrayBuffer(44 + dataSize);
  const dataView = new DataView(buffer);

  let pos = 0;
  const writeString = (value: string) => {
    for (let i = 0; i < value.length; i++) {
      dataView.setUint8(pos, value.charCodeAt(i));
      pos += 1;
    }
  };
  const writeUint32 = (value: number) => {
    dataView.setUint32(pos, value, true);
    pos += 4;
  };
  const writeUint16 = (value: number) => {
    dataView.setUint16(pos, value, true);
    pos += 2;
  };
  const writeSample = (offset: number, value: number) => {
    dataView.setFloat32(pos + offset * 4, value, true);
  };

  writeString("RIFF");
  writeUint32(36 + dataSize); // RIFFチャンクサイズ
  writeString("WAVE");
  writeString("fmt ");
  writeUint32(16); // fmtチャンクサイズ
  writeUint16(formatCode);
  writeUint16(numberOfChannels);
  writeUint32(sampleRate);
  writeUint32(byteRate);
  writeUint16(blockSize);
  writeUint16(bytesPerSample * 8); // 1サンプルあたりのビット数
  writeString("data");
  writeUint32(dataSize);

  for (let i = 0; i < numberOfChannels; i++) {
    const channelData = audioBuffer.getChannelData(i);
    for (let j = 0; j < numberOfSamples; j++) {
      writeSample(j * numberOfChannels + i, channelData[j]);
    }
  }

  return new Uint8Array(buffer);
};
