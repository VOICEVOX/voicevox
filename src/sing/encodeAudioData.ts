import {
  createMp3Encoder,
  createOggEncoder,
  WasmMediaEncoder,
} from "wasm-media-encoders";

export type SupportedAudioFormat = "wav" | "mp3" | "ogg";

const convertToWavFileData = (audioBuffer: AudioBuffer) => {
  const bytesPerSample = 4; // Float32
  const formatCode = 3; // WAVE_FORMAT_IEEE_FLOAT

  const numberOfChannels = audioBuffer.numberOfChannels;
  const numberOfSamples = audioBuffer.length;
  const sampleRate = audioBuffer.sampleRate;
  const byteRate = sampleRate * numberOfChannels * bytesPerSample;
  const blockSize = numberOfChannels * bytesPerSample;
  const dataSize = numberOfSamples * numberOfChannels * bytesPerSample;

  const buffer = new ArrayBuffer(44 + dataSize);
  const dataView = new DataView(buffer);

  let pos = 0;
  const writeString = (value: string) => {
    for (let i = 0; i < value.length; i++) {
      dataView.setUint8(pos, value.charCodeAt(i));
      pos += 1;
    }
  };
  const writeUint32 = (value: number) => {
    dataView.setUint32(pos, value, true);
    pos += 4;
  };
  const writeUint16 = (value: number) => {
    dataView.setUint16(pos, value, true);
    pos += 2;
  };
  const writeSample = (offset: number, value: number) => {
    dataView.setFloat32(pos + offset * 4, value, true);
  };

  writeString("RIFF");
  writeUint32(36 + dataSize); // RIFFチャンクサイズ
  writeString("WAVE");
  writeString("fmt ");
  writeUint32(16); // fmtチャンクサイズ
  writeUint16(formatCode);
  writeUint16(numberOfChannels);
  writeUint32(sampleRate);
  writeUint32(byteRate);
  writeUint16(blockSize);
  writeUint16(bytesPerSample * 8); // 1サンプルあたりのビット数
  writeString("data");
  writeUint32(dataSize);

  for (let i = 0; i < numberOfChannels; i++) {
    const channelData = audioBuffer.getChannelData(i);
    for (let j = 0; j < numberOfSamples; j++) {
      writeSample(j * numberOfChannels + i, channelData[j]);
    }
  }

  return new Uint8Array(buffer);
};

export const encodeAudioData = async (
  audioBuffer: AudioBuffer,
  encoder: WasmMediaEncoder<"audio/ogg" | "audio/mpeg">,
) => {
  let outBuffer = new Uint8Array(1024 * 1024);
  let offset = 0;
  let moreData = true;
  encoder.configure({
    channels: audioBuffer.numberOfChannels as 1 | 2,
    sampleRate: audioBuffer.sampleRate,
  });

  while (true) {
    const mp3Data = moreData
      ? encoder.encode(
          audioBuffer.numberOfChannels === 1
            ? [audioBuffer.getChannelData(0)]
            : [audioBuffer.getChannelData(0), audioBuffer.getChannelData(1)],
        )
      : encoder.finalize();

    if (mp3Data.length + offset > outBuffer.length) {
      const newBuffer = new Uint8Array(mp3Data.length + offset);
      newBuffer.set(outBuffer);
      outBuffer = newBuffer;
    }

    outBuffer.set(mp3Data, offset);
    offset += mp3Data.length;

    if (!moreData) {
      break;
    }

    moreData = false;
  }

  return outBuffer.slice(0, offset);
};

const convertToMp3Data = async (audioBuffer: AudioBuffer) => {
  const encoder = await createMp3Encoder();
  const mp3Data = await encodeAudioData(audioBuffer, encoder);
  return mp3Data;
};

const convertToOggData = async (audioBuffer: AudioBuffer) => {
  const encoder = await createOggEncoder();
  const oggData = await encodeAudioData(audioBuffer, encoder);
  return oggData;
};

export const convertToSupportedAudioFormat = async (
  audioBuffer: AudioBuffer,
  format: SupportedAudioFormat,
) => {
  switch (format) {
    case "wav":
      return convertToWavFileData(audioBuffer);
    case "mp3":
      return convertToMp3Data(audioBuffer);
    case "ogg":
      return convertToOggData(audioBuffer);
  }
};
