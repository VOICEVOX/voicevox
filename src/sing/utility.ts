export function round(value: number, digits: number) {
  const powerOf10 = 10 ** digits;
  return Math.round(value * powerOf10) / powerOf10;
}

export function linearInterpolation(
  x1: number,
  y1: number,
  x2: number,
  y2: number,
  x: number,
) {
  if (x2 <= x1) {
    throw new Error("x2 must be greater than x1.");
  }
  return y1 + ((y2 - y1) * (x - x1)) / (x2 - x1);
}

function ceilToOdd(value: number) {
  return 1 + Math.ceil((value - 1) / 2) * 2;
}

function createGaussianKernel(sigma: number) {
  const kernelSize = ceilToOdd(sigma * 3);
  const center = Math.floor(kernelSize / 2);
  let kernel: number[] = [];
  let sum = 0;
  for (let i = 0; i < kernelSize; i++) {
    const x = Math.abs(center - i);
    const value = Math.exp(-(x ** 2) / (2 * sigma ** 2));
    kernel.push(value);
    sum += value;
  }
  kernel = kernel.map((value) => value / sum);
  return kernel;
}

export function applyGaussianFilter(data: number[], sigma: number) {
  const kernel = createGaussianKernel(sigma);
  const center = Math.floor(kernel.length / 2);
  for (let i = 0; i < data.length; i++) {
    let sum = 0;
    for (let j = 0; j < kernel.length; j++) {
      let indexToRead = i - center + j;
      indexToRead = Math.max(0, indexToRead);
      indexToRead = Math.min(data.length - 1, indexToRead);
      sum += data[indexToRead] * kernel[j];
    }
    data[i] = sum;
  }
}

export async function calculateHash<T>(obj: T) {
  const textEncoder = new TextEncoder();
  const data = textEncoder.encode(JSON.stringify(obj));
  const digest = await crypto.subtle.digest("SHA-256", data);
  return Array.from(new Uint8Array(digest))
    .map((v) => v.toString(16).padStart(2, "0"))
    .join("");
}

export function createPromiseThatResolvesWhen(
  condition: () => boolean,
  interval = 200,
) {
  return new Promise<void>((resolve) => {
    const checkCondition = () => {
      if (condition()) {
        resolve();
      }
      window.setTimeout(checkCondition, interval);
    };
    checkCondition();
  });
}

/**
 * タイマーです。関数を定期的に実行します。
 */
export class Timer {
  private readonly interval: number;
  private timeoutId?: number;

  get isStarted() {
    return this.timeoutId != undefined;
  }

  /**
   * @param interval 関数を実行する間隔（ミリ秒）
   */
  constructor(interval: number) {
    this.interval = interval;
  }

  start(onTick: () => void) {
    const callback = () => {
      onTick();
      this.timeoutId = window.setTimeout(callback, this.interval);
    };
    this.timeoutId = window.setTimeout(callback, this.interval);
  }

  stop() {
    if (this.timeoutId == undefined) {
      throw new Error("The timer is not started.");
    }
    window.clearTimeout(this.timeoutId);
    this.timeoutId = undefined;
  }
}

/**
 * requestAnimationFrameを使用して関数を定期的に実行します。
 * 関数は、指定された最大フレームレート以下で実行されます。
 */
export class AnimationTimer {
  private readonly maxFrameTime: number;
  private readonly maxDiff: number;

  private requestId?: number;
  private prevTimeStamp?: number;
  private diff = 0;

  get isStarted() {
    return this.requestId != undefined;
  }

  /**
   * @param maxFrameRate 最大フレームレート（フレーム毎秒）
   */
  constructor(maxFrameRate = 60) {
    this.maxFrameTime = 1000 / maxFrameRate;
    this.maxDiff = this.maxFrameTime * 10;
  }

  start(onAnimationFrame: () => void) {
    if (this.requestId != undefined) {
      throw new Error("The animation frame runner is already started.");
    }

    this.diff = 0;
    this.prevTimeStamp = undefined;

    const callback = (timeStamp: number) => {
      if (this.prevTimeStamp == undefined) {
        this.diff += this.maxFrameTime;
      } else {
        this.diff += timeStamp - this.prevTimeStamp;
      }
      this.diff = Math.min(this.maxDiff, this.diff);
      if (this.diff >= this.maxFrameTime) {
        this.diff -= this.maxFrameTime;
        onAnimationFrame();
      }
      this.prevTimeStamp = timeStamp;
      this.requestId = window.requestAnimationFrame(callback);
    };
    this.requestId = window.requestAnimationFrame(callback);
  }

  stop() {
    if (this.requestId == undefined) {
      throw new Error("The animation frame runner is not started.");
    }
    window.cancelAnimationFrame(this.requestId);
    this.requestId = undefined;
  }
}
