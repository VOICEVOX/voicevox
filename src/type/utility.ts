// XとYが同じ型かどうかを判定する
export type IsEqual<X, Y> =
  (<T>() => T extends X ? 1 : 2) extends <T>() => T extends Y ? 1 : 2
    ? true
    : false;

// undefinedかnullでないことを保証する
export function assertNonNullable<T>(
  value: T,
): asserts value is NonNullable<T> {
  if (value == undefined) {
    throw new Error("Value is null or undefined");
  }
}

/**
 * never型になるはずの値を使って型システムで非到達をチェックできる便利クラス。
 * valueに指定した値がnever型じゃなかった場合に型システムがエラーを出してくれる。
 */
export class ExhaustiveError extends Error {
  constructor(value: never) {
    super(`Not exhaustive. value: ${value}`);
  }
}
