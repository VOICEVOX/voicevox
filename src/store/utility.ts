import { State } from "@/store/type";
import { ToolbarButtonTagType } from "@/type/preload";

export function sanitizeFileName(fileName: string): string {
  // \x00 - \x1f: ASCII 制御文字
  //   \x00: Null
  //   ...
  //   \x1f: Unit separator
  // \x22: "
  // \x2a: *
  // \x2f: /
  // \x3a: :
  // \x3c: <
  // \x3e: >
  // \x3f: ?
  // \x5c: \
  // \x7c: |
  // \x7f: DEL

  // eslint-disable-next-line no-control-regex
  const sanitizer = /[\x00-\x1f\x22\x2a\x2f\x3a\x3c\x3e\x3f\x5c\x7c\x7f]/g;

  return fileName.replace(sanitizer, "");
}

export function buildProjectFileName(state: State, extension?: string): string {
  const headItemText = state.audioItems[state.audioKeys[0]].text;

  const tailItemText =
    state.audioItems[state.audioKeys[state.audioKeys.length - 1]].text;

  const headTailItemText =
    headItemText !== tailItemText
      ? headItemText + "..." + tailItemText
      : headItemText;

  let defaultFileNameStem = sanitizeFileName(headTailItemText);

  if (defaultFileNameStem === "") {
    defaultFileNameStem = "Untitled";
  }

  return extension
    ? `${defaultFileNameStem}.${extension}`
    : defaultFileNameStem;
}

export const replaceTagIdToTagString = {
  index: "連番",
  characterName: "キャラ",
  styleName: "スタイル",
  text: "テキスト",
  date: "日付",
};
const replaceTagStringToTagId: { [tagString: string]: string } = Object.entries(
  replaceTagIdToTagString
).reduce((prev, [k, v]) => ({ ...prev, [v]: k }), {});

export const DEFAULT_FILE_NAME_TEMPLATE =
  "$連番$_$キャラ$（$スタイル$）_$テキスト$.wav";
const DEFAULT_FILE_NAME_VARIABLES = {
  index: 0,
  characterName: "四国めたん",
  text: "テキストテキストテキスト",
  styleName: "ノーマル",
  date: currentDateString(),
};

export function currentDateString(): string {
  const currentDate = new Date();
  const year = currentDate.getFullYear();
  const month = currentDate.getMonth().toString().padStart(2, "0");
  const date = currentDate.getDate().toString().padStart(2, "0");

  return `${year}${month}${date}`;
}

function replaceTag(
  template: string,
  replacer: { [key: string]: string }
): string {
  const result = template.replace(/\$(.+?)\$/g, (match, p1) => {
    const replaceTagId = replaceTagStringToTagId[p1];
    if (replaceTagId === undefined) {
      return match;
    }
    return replacer[replaceTagId] ?? "";
  });

  return result;
}

export function buildFileNameFromRawData(
  fileNamePattern = DEFAULT_FILE_NAME_TEMPLATE,
  vars = DEFAULT_FILE_NAME_VARIABLES
): string {
  let pattern = fileNamePattern;
  if (pattern === "") {
    // ファイル名指定のオプションが初期値("")ならデフォルトテンプレートを使う
    pattern = DEFAULT_FILE_NAME_TEMPLATE;
  }

  let text = sanitizeFileName(vars.text);
  if (text.length > 10) {
    text = text.substring(0, 9) + "…";
  }

  const characterName = sanitizeFileName(vars.characterName);

  const index = (vars.index + 1).toString().padStart(3, "0");

  const styleName = sanitizeFileName(vars.styleName);

  const date = currentDateString();

  return replaceTag(pattern, {
    index,
    characterName,
    styleName: styleName,
    text,
    date,
  });
}

export const getToolbarButtonName = (tag: ToolbarButtonTagType): string => {
  const tag2NameObj: Record<ToolbarButtonTagType, string> = {
    PLAY_CONTINUOUSLY: "連続再生",
    STOP: "停止",
    EXPORT_AUDIO_ONE: "１つ書き出し",
    EXPORT_AUDIO_ALL: "全部書き出し",
    EXPORT_AUDIO_CONNECT_ALL: "音声を繋げて書き出し",
    SAVE_PROJECT: "プロジェクト保存",
    UNDO: "元に戻す",
    REDO: "やり直す",
    IMPORT_TEXT: "テキスト読み込み",
    EMPTY: "空白",
  };
  return tag2NameObj[tag];
};

export const createKanaRegex = (includeSeparation?: boolean): RegExp => {
  // 以下の文字のみで構成される場合、「読み仮名」としてこれを処理する
  // includeSeparationがtrueの時は、読点(U+3001)とクエスチョン(U+FF1F)も含む
  //  * ひらがな(U+3041~U+3094)
  //  * カタカナ(U+30A1~U+30F4)
  //  * 全角長音(U+30FC)
  if (includeSeparation) {
    return /^[\u3041-\u3094\u30A1-\u30F4\u30FC\u3001\uFF1F]+$/;
  }
  return /^[\u3041-\u3094\u30A1-\u30F4\u30FC]+$/;
};

export const convertHiraToKana = (text: string): string => {
  return text.replace(/[\u3041-\u3094]/g, (s) => {
    return String.fromCharCode(s.charCodeAt(0) + 0x60);
  });
};

export const convertLongVowel = (text: string): string => {
  return text
    .replace(/(?<=[アカサタナハマヤラワャァガザダバパ]ー*)ー/g, "ア")
    .replace(/(?<=[イキシチニヒミリィギジヂビピ]ー*)ー/g, "イ")
    .replace(/(?<=[ウクスツヌフムユルュゥヴグズヅブプ]ー*)ー/g, "ウ")
    .replace(/(?<=[エケセテネヘメレェゲゼデベペ]ー*)ー/g, "エ")
    .replace(/(?<=[オコソトノホモヨロヲョォゴゾドボポ]ー*)ー/g, "オ")
    .replace(/(?<=[ン]ー*)ー/g, "ン")
    .replace(/(?<=[ッ]ー*)ー/g, "ッ");
};
