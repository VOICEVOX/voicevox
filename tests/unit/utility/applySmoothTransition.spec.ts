import { describe, it, expect } from "vitest";
import { applySmoothTransition } from "@/sing/utility";

describe("applySmoothTransition", () => {
  it("jumpIndices に非整数を含むと例外を投げる", () => {
    const data = [0, 1, 2];
    expect(() => applySmoothTransition(data, [1.5], 6)).toThrow();
  });

  it("jumpIndex が 0 以下または data.length 以上だと例外を投げる", () => {
    const data = [0, 1, 2, 3];
    expect(() => applySmoothTransition([...data], [0], 6)).toThrow();
    expect(() => applySmoothTransition([...data], [data.length], 6)).toThrow();
  });

  it("jumpIndices が昇順でないと例外を投げる", () => {
    const data = [0, 1, 2, 3, 4, 5];
    expect(() => applySmoothTransition(data, [3, 1], 6)).toThrow();
  });

  it("jumpIndices に重複があると例外を投げる", () => {
    const data = [0, 1, 2, 3];
    expect(() => applySmoothTransition(data, [2, 2], 6)).toThrow();
  });

  it("maxTransitionLength が有限でないと例外を投げる", () => {
    const data = [0, 1, 2, 3];
    expect(() => applySmoothTransition(data, [2], NaN)).toThrow();
  });

  it("maxTransitionLength < 2 のとき例外を投げる", () => {
    const data = [0, 1, 2, 3];
    expect(() => applySmoothTransition(data, [2], 1.5)).toThrow();
  });

  it("jumpIndices が空ならデータは変更されない", () => {
    const data = [0, 1, 2, 3, 4];
    const before = [...data];

    applySmoothTransition(data, [], 6);

    for (let i = 0; i < data.length; i++) {
      expect(data[i]).toBe(before[i]);
    }
  });

  it("中央付近の単一ジャンプに対して、滑らかな遷移が適用される", () => {
    const data = [0, 0, 0, 0, 1, 1, 1, 1, 1];
    const jumpIndices = [4];
    const maxTransitionLength = 6;

    const actual = [...data];
    applySmoothTransition(actual, jumpIndices, maxTransitionLength);

    expect(actual).toMatchSnapshot();
  });

  it("先頭近くのジャンプでは、遷移がデータの端に合わせて切り詰められる", () => {
    const data = [0, 1, 1, 1, 1];
    const jumpIndices = [1];
    const maxTransitionLength = 6;

    const actual = [...data];
    applySmoothTransition(actual, jumpIndices, maxTransitionLength);

    expect(actual).toMatchSnapshot();
  });

  it("末尾近くのジャンプでは、遷移がデータの端に合わせて切り詰められる", () => {
    const data = [0, 0, 0, 0, 1];
    const jumpIndices = [4];
    const maxTransitionLength = 6;

    const actual = [...data];
    applySmoothTransition(actual, jumpIndices, maxTransitionLength);

    expect(actual).toMatchSnapshot();
  });

  it("十分に離れた複数のジャンプに対して、それぞれ独立に遷移が適用される", () => {
    const data = [0, 0, 0, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2];
    const jumpIndices = [3, 9];
    const maxTransitionLength = 6;

    const actual = [...data];
    applySmoothTransition(actual, jumpIndices, maxTransitionLength);

    expect(actual).toMatchSnapshot();
  });

  it("隣接するジャンプでは遷移長が自動的に縮小され重複が避けられる", () => {
    const data = [0, 0, 0, 1, 1, 1, 2, 2, 2];
    const jumpIndices = [3, 6];
    const maxTransitionLength = 6;

    const actual = [...data];
    applySmoothTransition(actual, jumpIndices, maxTransitionLength);

    expect(actual).toMatchSnapshot();
  });

  it("ジャンプサイズが0の場合は、データは変更されない", () => {
    const data = [0, 0, 0, 0, 0];
    const actual = [...data];

    applySmoothTransition(actual, [3], 6);

    for (let i = 0; i < actual.length; i++) {
      expect(actual[i]).toBe(data[i]);
    }
  });
});
