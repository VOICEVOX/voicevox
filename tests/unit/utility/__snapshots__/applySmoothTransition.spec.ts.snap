// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`applySmoothTransition > 中央付近の単一ジャンプに対して、滑らかな遷移が適用される 1`] = `
[
  0,
  0.019675925925925927,
  0.15625,
  0.37615740740740744,
  0.6238425925925927,
  0.84375,
  0.9803240740740741,
  1,
  1,
]
`;

exports[`applySmoothTransition > 先頭近くのジャンプでは、遷移がデータの端に合わせて切り詰められる 1`] = `
[
  0.15625,
  0.84375,
  1,
  1,
  1,
]
`;

exports[`applySmoothTransition > 十分に離れた複数のジャンプに対して、それぞれ独立に遷移が適用される 1`] = `
[
  0.019675925925925927,
  0.15625,
  0.37615740740740744,
  0.6238425925925927,
  0.84375,
  0.9803240740740741,
  1.0196759259259258,
  1.15625,
  1.3761574074074074,
  1.6238425925925926,
  1.84375,
  1.980324074074074,
  2,
]
`;

exports[`applySmoothTransition > 末尾近くのジャンプでは、遷移がデータの端に合わせて切り詰められる 1`] = `
[
  0,
  0,
  0,
  0.15625,
  0.84375,
]
`;

exports[`applySmoothTransition > 隣接するジャンプでは遷移長が自動的に縮小され重複が避けられる 1`] = `
[
  0.019675925925925927,
  0.15625,
  0.37615740740740744,
  0.6439056605795611,
  1.003074363425926,
  1.363882726766118,
  1.6164413473079562,
  1.840675636574074,
  1.9799369320130316,
]
`;
